src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -435,18 +436,33 @@ ObjectDataHandler::write_ret do_remappings(
   return trans_intr::do_for_each(
     to_remap,
     [ctx](auto &region) {
+      auto fut = ObjectDataHandler::write_iertr::now();
+      if ((region.is_remap1() || region.is_remap2()) &&
+	  (!region.pin->is_parent_valid())) {
```

Comment:
But I'm still now sure whether `is_parent_valid()` is sufficient here. Seems to me if `pin` is retrived from a stable parent to a stable child, this cannot detect whether the parent has been modified, because the shared parent can still be valid. Is it correct?

Reply:
> But I'm still now sure whether `is_parent_valid()` is sufficient here. Seems to me if `pin` is retrived from a stable parent to a stable child, this cannot detect whether the parent has been modified, because the shared parent can still be valid. Is it correct?

You're right. I think the reason that we have met this issue is that, fortunately, all mappings are retrieved from mutable lba leaf nodes when we want to `get pending logical extents` through them.

On the other hand, wrongly getting pending logical extents through stable mappings happens when those stable mappings have been updated, I think maybe we can implement the `copy_sources -> initial_pending` links mentioned in https://github.com/ceph/ceph/pull/55831#issuecomment-1990898098, so that we can check whether the stable mappings have been modified when NDEBUG is not defined, which should be able to prevent this issue. What do you think?

Reply:
Probably a simpler fix for now could be to pass the transaction reference to `is_parent_valid()`, because if the stable parent was modified, it must be retired in that transaction. What do you think?

Reply:
> Probably a simpler fix for now could be to pass the transaction reference to `is_parent_valid()`, because if the stable parent was modified, it must be retired in that transaction. What do you think?

I think if the parent has been modified while the mapping hasn't been updated, that mapping should be considered valid for getting logical extents. Because, in this scenario, even if we try to get the logical extents from mutated lba leaf nodes, we still have to fall back to the stable copy sources to find the lba pointer. What do you think?

Reply:
I see, I think you are right.

IIUC the correctness of `is_parent_valid()` is under the assumption that the child (ObjectDataHandler, i.e. 
`region.pin`) is stable upon remapping. So that the lookup is still correct through the stable parent even if the parent has been invalidated (mutated or splitted) in the transaction.

If that's correct, how about adding `ceph_assert(region.pin->is_stable())` with some comments to explain ? (It's not very easy to follow because it leverages low-level properties of the lba pointer design.

Reply:
Also, a related unit test case is needed to prove that this keeps working when there is a related lba modification happened after getting the original pin. (Can we also validate that the original pin is wrong in this case?)

Reply:
@cyx1231st I added checks for debug builds to make sure that lba mappings haven't been modified when being traced to get logical extents. This can accurately capture the scenarios that would lead to errors, and as long as these checks are passed, there would be no problem. Please take a look, thanks.